File: rust/crates/lfs_frb/src/api/local_fs.rs
//! FRB adapter for `lfs_core::fs::local`. The Dart `LocalFS`
//! file_browser implementation routes every operation here so
//! `dart:io File / Directory` no longer participates in the
//! file_browser path. `initialDir` stays Dart-side because it
//! depends on `path_provider` (iOS sandbox / Android scoped
//! storage) which has no clean Rust analog.

#[derive(Debug, Clone)]
pub struct DbLocalFileEntry {
    pub name: String,
    pub path: String,
    pub size: u64,
    pub mode: u32,
    pub mod_time_unix_ms: i64,
    pub is_dir: bool,
    /// `true` only when the entry was produced by
    /// [`local_fs_symlink_stat`] and the path is itself a
    /// symbolic link. List / stat results report `false`.
    pub is_symlink: bool,
}

fn to_db_entry(e: lfs_core::fs::local::LocalFileEntry) -> DbLocalFileEntry {
    DbLocalFileEntry {
        name: e.name,
        path: e.path,
        size: e.size,
        mode: e.mode,
        mod_time_unix_ms: e.mod_time_unix_ms,
        is_dir: e.is_dir,
        is_symlink: e.is_symlink,
    }
}

pub async fn local_fs_list(path: String) -> Result<Vec<DbLocalFileEntry>, String> {
    Ok(lfs_core::fs::local::list(path)
        .await?
        .into_iter()
        .map(to_db_entry)
        .collect())
}

/// Stat `path` following symlinks. `None` means "does not
/// exist"; other I/O failures (permission denied, broken disk)
/// surface as `Err`. The transfer walker uses this to read size
/// and mtime for the change-detected replace path; the file pane
/// uses it as a one-trip "exists?" probe.
pub async fn local_fs_stat(path: String) -> Result<Option<DbLocalFileEntry>, String> {
    Ok(lfs_core::fs::local::stat(path).await?.map(to_db_entry))
}

/// Stat `path` without following symlinks. The returned
/// entry's `is_symlink` is `true` when the path itself is a
/// symlink, regardless of target type — matches Dart's
/// `FileSystemEntity.typeSync(path, followLinks: false)`.
pub async fn local_fs_symlink_stat(path: String) -> Result<Option<DbLocalFileEntry>, String> {
    Ok(lfs_core::fs::local::symlink_stat(path)
        .await?
        .map(to_db_entry))
}

pub async fn local_fs_list_directories(path: String) -> Result<Vec<String>, String> {
    lfs_core::fs::local::list_directories(path).await
}

pub async fn local_fs_mkdir(path: String) -> Result<(), String> {
    lfs_core::fs::local::mkdir(path).await
}

pub async fn local_fs_remove(path: String) -> Result<(), String> {
    lfs_core::fs::local::remove(path).await
}

pub async fn local_fs_remove_dir(path: String) -> Result<(), String> {
    lfs_core::fs::local::remove_dir(path).await
}

pub async fn local_fs_rename(old_path: String, new_path: String) -> Result<(), String> {
    lfs_core::fs::local::rename(old_path, new_path).await
}

pub async fn local_fs_dir_size(path: String) -> Result<u64, String> {
    lfs_core::fs::local::dir_size(path).await
}

pub async fn local_fs_windows_hidden_names(dir: String) -> Vec<String> {
    lfs_core::fs::local::windows_hidden_names(dir).await
}

/// Copy a single file. Replaces the destination if it exists.
/// Symmetric with [`local_fs_copy_recursive_no_symlinks`] so the
/// file-browser drop path stays Rust-side across both branches.
pub async fn local_fs_copy_file(src: String, dst: String) -> Result<(), String> {
    lfs_core::fs::local::copy_file(src, dst).await
}

/// Recursively copy a directory tree. Refuses to traverse
/// symlinks: a symlink at the root returns
/// `Err("symlink_in_source")`, symlinks inside the tree are
/// silently skipped, and recursion is bounded by `max_depth`.
/// The Dart file-browser drop path passes 100, matching the
/// constant it previously enforced inline.
pub async fn local_fs_copy_recursive_no_symlinks(
    src: String,
    dst: String,
    max_depth: u32,
) -> Result<(), String> {
    lfs_core::fs::local::copy_recursive_no_symlinks(src, dst, max_depth).await
}

/// Android shared-storage initial-dir probe. Stats
/// `/storage/emulated/0`, then stats and lists `Download` to
/// detect MANAGE_EXTERNAL_STORAGE (a permissive root listing
/// succeeds on scoped storage even when the child is locked).
/// `None` means the probe failed; the Dart caller falls back
/// to `getExternalStorageDirectory()` (Flutter plugin path,
/// no Rust analog). `home_dir` is the canonical shared root
/// (`/storage/emulated/0` on most devices).
pub async fn local_fs_android_initial_dir(home_dir: String) -> Option<String> {
    lfs_core::fs::local::android_initial_dir_probe(home_dir).await
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn list_empty_dir_returns_empty_vec() {
        let tmp = tempfile::tempdir().expect("tmp dir");
        let path = tmp.path().to_str().expect("utf-8 path").to_string();
        let entries = local_fs_list(path).await.expect("list empty");
        assert!(entries.is_empty());
    }

    #[tokio::test]
    async fn list_returns_entries_with_populated_fields() {
        let tmp = tempfile::tempdir().expect("tmp dir");
        std::fs::write(tmp.path().join("file.txt"), b"hello").expect("write");
        std::fs::create_dir(tmp.path().join("sub")).expect("mkdir");
        let path = tmp.path().to_str().expect("utf-8 path").to_string();
        let entries = local_fs_list(path).await.expect("list");
        assert_eq!(entries.len(), 2);
        let file = entries
            .iter()
            .find(|e| e.name == "file.txt")
            .expect("file row");
        assert!(!file.is_dir);
        assert_eq!(file.size, 5);
        let sub = entries.iter().find(|e| e.name == "sub").expect("dir row");
        assert!(sub.is_dir);
    }

    #[tokio::test]
    async fn mkdir_creates_a_new_directory() {
        let tmp = tempfile::tempdir().expect("tmp dir");
        let target = tmp.path().join("new_subdir");
        let target_str = target.to_str().expect("utf-8 path").to_string();
        local_fs_mkdir(target_str).await.expect("mkdir");
        assert!(target.is_dir());
    }

    #[tokio::test]
    async fn remove_deletes_a_file() {
        let tmp = tempfile::tempdir().expect("tmp dir");
        let file = tmp.path().join("delete-me.txt");
        std::fs::write(&file, b"x").expect("write");
        let file_str = file.to_str().expect("utf-8 path").to_string();
        local_fs_remove(file_str).await.expect("remove");
        assert!(!file.exists());
    }

    #[tokio::test]
    async fn remove_dir_drops_empty_directory() {
        let tmp = tempfile::tempdir().expect("tmp dir");
        let dir = tmp.path().join("empty-subdir");
        std::fs::create_dir(&dir).expect("mkdir");
        let dir_str = dir.to_str().expect("utf-8 path").to_string();
        local_fs_remove_dir(dir_str).await.expect("rmdir");
        assert!(!dir.exists());
    }

    #[tokio::test]
    async fn rename_relocates_a_file() {
        let tmp = tempfile::tempdir().expect("tmp dir");
        let src = tmp.path().join("src.txt");
        let dst = tmp.path().join("dst.txt");
        std::fs::write(&src, b"hello").expect("write");
        local_fs_rename(
            src.to_str().expect("utf-8").to_string(),
            dst.to_str().expect("utf-8").to_string(),
        )
        .await
        .expect("rename");
        assert!(!src.exists());
        assert!(dst.exists());
        assert_eq!(std::fs::read(&dst).expect("read"), b"hello");
    }

    #[tokio::test]
    async fn dir_size_sums_recursive_file_sizes() {
        let tmp = tempfile::tempdir().expect("tmp dir");
        std::fs::write(tmp.path().join("a.bin"), [0u8; 100]).expect("write a");
        std::fs::write(tmp.path().join("b.bin"), [0u8; 200]).expect("write b");
        std::fs::create_dir(tmp.path().join("sub")).expect("mkdir");
        std::fs::write(tmp.path().join("sub/c.bin"), [0u8; 50]).expect("write c");
        let path = tmp.path().to_str().expect("utf-8 path").to_string();
        let total = local_fs_dir_size(path).await.expect("dir_size");
        assert_eq!(total, 350);
    }
}

File: lib/platform/local_fs.dart
import 'dart:io';

import 'package:meta/meta.dart' show visibleForTesting;
import 'package:path_provider/path_provider.dart';

import '../core/sftp/file_system.dart';
import '../core/sftp/sftp_models.dart';
import '../src/rust/api/local_fs.dart' as rust_local_fs;
import '../utils/logger.dart';
import '../utils/platform.dart';

/// Local file system implementation.
///
/// File operations route through `lfs_core::fs::local` (FRB async),
/// but the adapter lives in `platform/` rather than `core/` because
/// [initialDir] resolves the OS sandbox start directory via
/// `path_provider` (iOS Documents, Android scoped storage) — a
/// Flutter-plugin path with no clean Rust analog. The network
/// `FileSystem` implementations (SftpFS / WebdavFS) stay in `core/`;
/// only the local one is platform-bound.
class LocalFS implements FileSystem {
  @override
  Future<String> initialDir() async {
    // iOS sandbox: start in the app's Documents folder (visible in Files.app).
    // Users can pick external folders via the folder picker button.
    if (Platform.isIOS) {
      final docs = await getApplicationDocumentsDirectory();
      AppLogger.instance.log('iOS initial dir: <path>', name: 'LocalFS');
      return docs.path;
    }
    if (Platform.isAndroid) {
      return _androidInitialDir();
    }
    final home = homeDirectory;
    final path = home.isNotEmpty ? home : Directory.current.path;
    AppLogger.instance.log('Initial dir: <path>', name: 'LocalFS');
    return path;
  }

  /// Android: try shared storage first, fall back to app-specific dir
  /// if storage permission is not granted.
  ///
  /// The probe runs Rust-side via `localFsAndroidInitialDir` —
  /// it stats `/storage/emulated/0`, then stats and lists the
  /// `Download` subdirectory because scoped storage on Android
  /// 11+ lets apps see folder names at the root without actual
  /// read access. `null` here means the probe failed (permission
  /// denied / not mounted) and the Dart caller pivots to
  /// `getExternalStorageDirectory()` (Flutter plugin path, no
  /// Rust analog) for the app-specific fallback.
  Future<String> _androidInitialDir() async {
    final probed = await rust_local_fs.localFsAndroidInitialDir(
      homeDir: homeDirectory,
    );
    if (probed != null) return probed;
    AppLogger.instance.log(
      'No shared storage access, falling back to app dir',
      name: 'LocalFS',
      level: LogLevel.warn,
    );
    final appDir = await getExternalStorageDirectory();
    final fallbackPath = appDir?.path ?? Directory.current.path;
    AppLogger.instance.log('Android fallback dir: <path>', name: 'LocalFS');
    return fallbackPath;
  }

  @override
  Future<List<FileEntry>> list(String path) async {
    final List<rust_local_fs.DbLocalFileEntry> rows;
    try {
      // `localFsListVisible` drops Windows Hidden / System files
      // Rust-side so the pane matches Explorer — the hidden-name
      // decision + filter loop are no longer Dart's. Identical to
      // a plain list on every non-Windows target.
      rows = await rust_local_fs.localFsListVisible(path: path);
    } catch (e) {
      // Re-throw as FileSystemException so callers that catch it
      // see one stable exception type regardless of FRB error
      // variant. FRB surfaces `Result<_, String>` errors as either
      // an `AnyhowException` or a plain `String`-bearing exception
      // depending on codec; a single broad catch unifies both.
      throw FileSystemException(describeError(e), path);
    }

    final entries = <FileEntry>[];
    for (final row in rows) {
      entries.add(
        FileEntry(
          name: row.name,
          path: row.path,
          size: row.size.toInt(),
          mode: row.mode,
          modTime: DateTime.fromMillisecondsSinceEpoch(
            row.modTimeUnixMs.toInt(),
          ),
          isDir: row.isDir,
          isSymlink: row.isSymlink,
        ),
      );
    }
    sortFileEntries(entries);
    return entries;
  }

  @override
  Future<void> mkdir(String path) async {
    AppLogger.instance.log('Creating directory: <path>', name: 'LocalFS');
    await rust_local_fs.localFsMkdir(path: path);
  }

  @override
  Future<void> remove(String path) async {
    AppLogger.instance.log('Removing: <path>', name: 'LocalFS');
    await rust_local_fs.localFsRemove(path: path);
  }

  @override
  Future<void> removeDir(String path) async {
    AppLogger.instance.log(
      'Removing directory recursively: <path>',
      name: 'LocalFS',
    );
    await rust_local_fs.localFsRemoveDir(path: path);
  }

  @override
  Future<int> dirSize(String path) async {
    final size = await rust_local_fs.localFsDirSize(path: path);
    return size.toInt();
  }

  /// Single FRB call — the recursive enumeration, symlink-skip, and
  /// per-segment name validation all run in `lfs_core::fs::local`.
  /// The upload walker calls this instead of recursing through
  /// `dart:io` / per-level `list`.
  @override
  Future<List<FlatFileLeaf>> flatWalkFiles(
    String root, {
    int maxDepth = 100,
  }) async {
    final leaves = await rust_local_fs.localFsFlatWalkFiles(
      root: root,
      maxDepth: maxDepth,
    );
    return [
      for (final e in leaves)
        FlatFileLeaf(relPath: e.relPath, size: e.size.toInt()),
    ];
  }

  /// Cheap presence probe — `localFsSymlinkStat` returns `null`
  /// when the path doesn't exist and a populated record otherwise,
  /// without following symlinks. The trait's default parent-list
  /// fallback would burn a full directory listing per probe; the
  /// symlink-aware stat is one syscall.
  @override
  Future<bool> exists(String path) async {
    try {
      return await rust_local_fs.localFsSymlinkStat(path: path) != null;
    } catch (_) {
      return false;
    }
  }

  /// LocalFS lists carry `st_mode` (Unix) or a synthesised mode
  /// (Windows / Android) and `owner` (uid name on Unix, account
  /// string on Windows) on every entry, so both columns always
  /// have useful content.
  @override
  FileSystemCapabilities get capabilities => FileSystemCapabilities.posix;

  @override
  Future<void> rename(String oldPath, String newPath) async {
    AppLogger.instance.log('Renaming: <path> → <path>', name: 'LocalFS');
    await rust_local_fs.localFsRename(oldPath: oldPath, newPath: newPath);
  }

  /// Pull a one-line message out of the FRB error envelope so the
  /// `FileSystemException("Directory not found", path)` shape stays
  /// human-readable in the file_browser error toasts.
  @visibleForTesting
  static String describeError(Object e) {
    final s = e.toString();
    // FRB throws a wrapper class whose toString includes the full
    // message we passed back from Rust; trim a leading
    // "AnyhowException: " when present so the toast doesn't show it.
    const prefix = 'AnyhowException: ';
    if (s.startsWith(prefix)) return s.substring(prefix.length);
    return s;
  }
}
